In [18]:
# ============================================================
# STEP 4 — SCRAPE HTML TABLES USING pd.read_html()
# ============================================================

import pandas as pd
import requests

# Source website
url = "https://en.wikipedia.org/wiki/Shark_Tank_India_season_1"

# Browser-like headers to avoid HTTP 403
headers = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/153.0.0.0 Safari/537.36"
    )
}

# Fetch webpage
response = requests.get(url, headers=headers, timeout=30)

print("Status Code:", response.status_code)

# Check whether webpage was successfully fetched
response.raise_for_status()

# Scrape all HTML tables using pandas
tables = pd.read_html(response.text)

print("Number of tables scraped:", len(tables))

# Display basic information about scraped tables
for i, table in enumerate(tables):
    print(f"\nTable {i}: {table.shape}")
    display(table.head(3))


Status Code: 200
Number of tables scraped: 5

Table 0: (7, 2)


/tmp/ipykernel_1245/302976268.py:29: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  tables = pd.read_html(response.text)


,Shark Tank India,Shark Tank India.1
0,Season 1,Season 1
1,Starring,Ashneer Grover Aman Gupta Anupam Mittal Ghazal...
2,No. of episodes,36



Table 1: (7, 2)


,Shark,Company
0,Ashneer Grover,ex-Managing Director and co-founder of BharatPe
1,Aman Gupta,Co-founder and Chief Marketing Officer of boAt...
2,Anupam Mittal,Founder and CEO of Shaadi.com and People Group



Table 2: (124, 15)


Ep. No.           Pitch No.l                Brand  \
  Ep. No. Ep. No..1 Pitch No.l                Brand   
0       1         1          1  BluePine Industries   
1       1         1          2        Booz scooters   
2       1         1          3  Heart up my Sleeves   

                                            Idea       City  \
                                            Idea       City   
0                                   Frozen Momos  Faridabad   
1  Renting e-bike for mobility in private spaces  Ahmedabad   
2                             Detachable Sleeves  New Delhi   

               Original ask                      Deal   Invested by  \
               Original ask                      Deal Anupam Mittal   
0   ₹50 lakhs for 5% equity  ₹75 lakhs for 16% equity           NaN   
1  ₹40 lakhs for 15% equity  ₹40 lakhs for 50% equity           NaN   
2  ₹25 lakhs for 10% equity  ₹25 lakhs for 30% equity           NaN   

                                                                       \
  Aman Gupta Ashneer Grover Namita Thapar Peyush Bansal Vineeta Singh   
0        NaN            NaN           NaN          —N/a           NaN   
1        NaN            NaN           NaN          —N/a           NaN   
2        NaN            NaN           NaN          —N/a           NaN   

                
  Ghazal Alagh  
0         —N/a  
1         —N/a  
2         —N/a


Table 3: (26, 2)


,Brand,Original ask
,Brand,Original ask
0,Ashwa Pro,50 lakhs for 10% equity
1,Canebot,1 crore for 2.5% equity
2,Clensta,1 crore for 1% equity



Table 4: (3, 2)


,vteCurrent broadcasts on Sony Entertainment Television,vteCurrent broadcasts on Sony Entertainment Television.1
0,Drama,"CID Chalo Bulawa Aya Hai, Mata Ne Bulaya Hai"
1,Reality/non-scripted series,India's Got Talent Kaun Banega Crorepati India...
2,List of former SET shows - List of Sony SAB Pr...,List of former SET shows - List of Sony SAB Pr...


## 5. Select the Required Dataset

The scraped webpage contains multiple HTML tables.

For this project, the main dataset is the **pitch and investment table**, which contains information about businesses presented on Shark Tank India, including their episode number, pitch number, brand, business idea, city, original funding ask, deal information, and shark investments.

The required table is selected from the scraped HTML tables for further data cleaning, transformation, and optimization.


In [19]:
# ============================================================
# STEP 5 — SELECT THE REQUIRED DATASET
# ============================================================

# The main pitch and investment table is Table 2
shark_tank_df = tables[2].copy()

# Display dataset dimensions
print("Dataset Shape:", shark_tank_df.shape)

# Display column names
print("\nColumn Names:")
print(shark_tank_df.columns.tolist())

# Display first 5 rows
print("\nFirst 5 Rows:")
display(shark_tank_df.head())

Dataset Shape: (124, 15)

Column Names:
[('Ep. No.', 'Ep. No.'), ('Ep. No.', 'Ep. No..1'), ('Pitch No.l', 'Pitch No.l'), ('Brand', 'Brand'), ('Idea', 'Idea'), ('City', 'City'), ('Original ask', 'Original ask'), ('Deal', 'Deal'), ('Invested by', 'Anupam Mittal'), ('Invested by', 'Aman Gupta'), ('Invested by', 'Ashneer Grover'), ('Invested by', 'Namita Thapar'), ('Invested by', 'Peyush Bansal'), ('Invested by', 'Vineeta Singh'), ('Invested by', 'Ghazal Alagh')]

First 5 Rows:


Ep. No.           Pitch No.l                Brand  \
  Ep. No. Ep. No..1 Pitch No.l                Brand   
0       1         1          1  BluePine Industries   
1       1         1          2        Booz scooters   
2       1         1          3  Heart up my Sleeves   
3       2         2          4           Tagz Foods   
4       2         2          5       Head and Heart   

                                            Idea       City  \
                                            Idea       City   
0                                   Frozen Momos  Faridabad   
1  Renting e-bike for mobility in private spaces  Ahmedabad   
2                             Detachable Sleeves  New Delhi   
3                           Healthy Potato Chips  Bengaluru   
4                       Brain Development Course        NaN   

               Original ask                        Deal   Invested by  \
               Original ask                        Deal Anupam Mittal   
0   ₹50 lakhs for 5% equity    ₹75 lakhs for 16% equity           NaN   
1  ₹40 lakhs for 15% equity    ₹40 lakhs for 50% equity           NaN   
2  ₹25 lakhs for 10% equity    ₹25 lakhs for 30% equity           NaN   
3   ₹70 lakhs for 1% equity  ₹70 lakhs for 2.75% equity           NaN   
4   ₹50 lakhs for 5% equity                     No Deal           NaN   

                                                                       \
  Aman Gupta Ashneer Grover Namita Thapar Peyush Bansal Vineeta Singh   
0        NaN            NaN           NaN          —N/a           NaN   
1        NaN            NaN           NaN          —N/a           NaN   
2        NaN            NaN           NaN          —N/a           NaN   
3        NaN            NaN           NaN          —N/a           NaN   
4        NaN            NaN           NaN          —N/a           NaN   

                
  Ghazal Alagh  
0         —N/a  
1         —N/a  
2         —N/a  
3         —N/a  
4         —N/a

## 6. Data Cleaning — Column Names and Repeated Headers

The scraped dataset contains multi-level column headers because the original webpage groups the shark investment columns under a common "Invested by" heading.

For better usability and analysis, the following cleaning operations are performed:

- Flatten the multi-level column headers.
- Retain meaningful shark names as individual columns.
- Remove unnecessary duplicate header information.
- Remove the repeated header row present inside the dataset.
- Reset the DataFrame index.
- Standardize column names for easier analysis.

In [20]:
# ============================================================
# STEP 6 — CLEAN COLUMN NAMES AND REMOVE REPEATED HEADERS
# ============================================================

# Create a copy so the scraped dataset remains unchanged
clean_df = shark_tank_df.copy()

# ------------------------------------------------------------
# 1. Flatten MultiIndex columns
# ------------------------------------------------------------

if isinstance(clean_df.columns, pd.MultiIndex):

    cleaned_columns = []

    for column in clean_df.columns:

        # Remove empty / NaN values
        parts = [
            str(part).strip()
            for part in column
            if pd.notna(part) and str(part).strip() != ""
        ]

        # Remove duplicate consecutive values
        unique_parts = []
        for part in parts:
            if part not in unique_parts:
                unique_parts.append(part)

        # For grouped columns such as "Invested by -> Aman Gupta",
        # keep the meaningful lower-level column name.
        if len(unique_parts) > 1:
            column_name = unique_parts[-1]
        elif len(unique_parts) == 1:
            column_name = unique_parts[0]
        else:
            column_name = "Unknown"

        cleaned_columns.append(column_name)

    clean_df.columns = cleaned_columns

else:
    # If columns are already single-level
    clean_df.columns = [
        str(column).strip()
        for column in clean_df.columns
    ]


# ------------------------------------------------------------
# 2. Standardize column names
# ------------------------------------------------------------

clean_df.columns = (
    clean_df.columns
    .str.replace(r"\s+", " ", regex=True)
    .str.strip()
)


# ------------------------------------------------------------
# 3. Remove repeated header row
# ------------------------------------------------------------

# The scraped table contains a repeated header row inside
# the actual dataset. Remove it using the Brand column.
if "Brand" in clean_df.columns:

    clean_df = clean_df[
        clean_df["Brand"].astype(str).str.strip().ne("Brand")
    ]


# ------------------------------------------------------------
# 4. Reset index
# ------------------------------------------------------------

clean_df = clean_df.reset_index(drop=True)


# ------------------------------------------------------------
# 5. Convert Episode and Pitch numbers to numeric
# ------------------------------------------------------------

for column in ["Ep. No.", "Pitch No."]:
    if column in clean_df.columns:
        clean_df[column] = pd.to_numeric(
            clean_df[column],
            errors="coerce"
        )


# ------------------------------------------------------------
# 6. Display cleaned dataset information
# ------------------------------------------------------------

print("Cleaned Dataset Shape:", clean_df.shape)

print("\nCleaned Column Names:")
print(clean_df.columns.tolist())

print("\nFirst 5 Rows:")
display(clean_df.head())

print("\nData Types:")
display(clean_df.dtypes)


Cleaned Dataset Shape: (124, 15)

Cleaned Column Names:
['Ep. No.', 'Ep. No..1', 'Pitch No.l', 'Brand', 'Idea', 'City', 'Original ask', 'Deal', 'Anupam Mittal', 'Aman Gupta', 'Ashneer Grover', 'Namita Thapar', 'Peyush Bansal', 'Vineeta Singh', 'Ghazal Alagh']

First 5 Rows:


,Ep. No.,Ep. No..1,Pitch No.l,Brand,Idea,City,Original ask,Deal,Anupam Mittal,Aman Gupta,Ashneer Grover,Namita Thapar,Peyush Bansal,Vineeta Singh,Ghazal Alagh
0,1.0,1,1,BluePine Industries,Frozen Momos,Faridabad,₹50 lakhs for 5% equity,₹75 lakhs for 16% equity,NaN,NaN,NaN,NaN,—N/a,NaN,—N/a
1,1.0,1,2,Booz scooters,Renting e-bike for mobility in private spaces,Ahmedabad,₹40 lakhs for 15% equity,₹40 lakhs for 50% equity,NaN,NaN,NaN,NaN,—N/a,NaN,—N/a
2,1.0,1,3,Heart up my Sleeves,Detachable Sleeves,New Delhi,₹25 lakhs for 10% equity,₹25 lakhs for 30% equity,NaN,NaN,NaN,NaN,—N/a,NaN,—N/a
3,2.0,2,4,Tagz Foods,Healthy Potato Chips,Bengaluru,₹70 lakhs for 1% equity,₹70 lakhs for 2.75% equity,NaN,NaN,NaN,NaN,—N/a,NaN,—N/a
4,2.0,2,5,Head and Heart,Brain Development Course,NaN,₹50 lakhs for 5% equity,No Deal,NaN,NaN,NaN,NaN,—N/a,NaN,—N/a



Data Types:


,0
Ep. No.,float64
Ep. No..1,object
Pitch No.l,object
Brand,object
Idea,object
City,object
Original ask,object
Deal,object
Anupam Mittal,object
Aman Gupta,object


## 7. Missing Values Analysis and Cleaning

Missing values are analyzed across all columns to understand the completeness of the scraped dataset.

The following steps are performed:

- Count missing values in each column.
- Calculate the percentage of missing values.
- Identify columns containing missing data.
- Determine an appropriate treatment based on the nature of the data.
- Replace missing categorical values with a consistent label where appropriate.

In [21]:
# ============================================================
# STEP 7 — MISSING VALUES ANALYSIS AND CLEANING
# ============================================================

# Create a missing-value summary
missing_summary = pd.DataFrame({
    "Missing Values": clean_df.isna().sum(),
    "Missing Percentage": (
        clean_df.isna().mean() * 100
    ).round(2)
})

# Keep only columns containing missing values
missing_summary = (
    missing_summary[
        missing_summary["Missing Values"] > 0
    ]
    .sort_values("Missing Values", ascending=False)
)

print("Missing Value Summary:")
display(missing_summary)

# ------------------------------------------------------------
# Replace missing categorical/text values
# ------------------------------------------------------------

# Shark investment columns
shark_columns = [
    "Anupam Mittal",
    "Aman Gupta",
    "Ashneer Grover",
    "Namita Thapar",
    "Peyush Bansal",
    "Vineeta Singh",
    "Ghazal Alagh"
]

# Replace missing values in shark investment columns
# with "No Investment Recorded"
for column in shark_columns:
    if column in clean_df.columns:
        clean_df[column] = clean_df[column].fillna(
            "No Investment Recorded"
        )

# Replace missing values in general categorical columns
categorical_columns = ["Brand", "Idea", "City", "Deal"]

for column in categorical_columns:
    if column in clean_df.columns:
        clean_df[column] = clean_df[column].fillna(
            "Not Available"
        )

# ------------------------------------------------------------
# Verify missing values after cleaning
# ------------------------------------------------------------

remaining_missing = clean_df.isna().sum()

print("\nRemaining Missing Values:")
display(
    remaining_missing[
        remaining_missing > 0
    ].to_frame("Missing Values")
)

print("\nDataset Shape After Missing Value Treatment:")
print(clean_df.shape)


Missing Value Summary:


,Missing Values,Missing Percentage
Anupam Mittal,121,97.58
Namita Thapar,110,88.71
Aman Gupta,102,82.26
City,101,81.45
Ashneer Grover,98,79.03
Peyush Bansal,92,74.19
Vineeta Singh,70,56.45
Ghazal Alagh,26,20.97
Ep. No.,22,17.74
Idea,4,3.23



Remaining Missing Values:


,Missing Values
Ep. No.,22



Dataset Shape After Missing Value Treatment:
(124, 15)


## 8. Duplicate Check and Data Type Optimization

The cleaned dataset is checked for duplicate records to avoid repeated observations.

The data types of important columns are also optimized to improve consistency and usability for further analysis.

The following operations are performed:

- Identify duplicate rows.
- Remove exact duplicate records, if present.
- Convert episode and pitch numbers to integer type.
- Convert appropriate text columns to string type.
- Verify the final data types and dataset structure.

In [22]:
# ============================================================
# STEP 8 — DUPLICATE CHECK AND DATA TYPE OPTIMIZATION
# ============================================================

# ------------------------------------------------------------
# 1. Check for duplicate rows
# ------------------------------------------------------------

duplicate_count = clean_df.duplicated().sum()

print("Number of Duplicate Rows:", duplicate_count)

# Remove exact duplicate rows, if any
if duplicate_count > 0:
    clean_df = clean_df.drop_duplicates().reset_index(drop=True)

print("Dataset Shape After Duplicate Removal:", clean_df.shape)


# ------------------------------------------------------------
# 2. Optimize numeric columns
# ------------------------------------------------------------

for column in ["Ep. No.", "Pitch No."]:
    if column in clean_df.columns:
        clean_df[column] = pd.to_numeric(
            clean_df[column],
            errors="coerce"
        ).astype("Int64")


# ------------------------------------------------------------
# 3. Standardize text columns
# ------------------------------------------------------------

text_columns = [
    "Brand",
    "Idea",
    "City",
    "Original ask",
    "Deal",
    "Anupam Mittal",
    "Aman Gupta",
    "Ashneer Grover",
    "Namita Thapar",
    "Peyush Bansal",
    "Vineeta Singh",
    "Ghazal Alagh"
]

for column in text_columns:
    if column in clean_df.columns:
        clean_df[column] = clean_df[column].astype("string").str.strip()


# ------------------------------------------------------------
# 4. Verify data types
# ------------------------------------------------------------

print("\nFinal Data Types:")
display(clean_df.dtypes.to_frame("Data Type"))


# ------------------------------------------------------------
# 5. Final dataset information
# ------------------------------------------------------------

print("\nDataset Information:")
print("Rows    :", clean_df.shape[0])
print("Columns :", clean_df.shape[1])

print("\nDuplicate Rows Remaining:", clean_df.duplicated().sum())

Number of Duplicate Rows: 0
Dataset Shape After Duplicate Removal: (124, 15)

Final Data Types:


,Data Type
Ep. No.,Int64
Ep. No..1,object
Pitch No.l,object
Brand,string[python]
Idea,string[python]
City,string[python]
Original ask,string[python]
Deal,string[python]
Anupam Mittal,string[python]
Aman Gupta,string[python]



Dataset Information:
Rows    : 124
Columns : 15

Duplicate Rows Remaining: 0


## 9. Data Validation and Quality Check

The cleaned dataset is validated to ensure that it is consistent, complete, and suitable for further analysis.

The following quality checks are performed:

- Verify the final number of rows and columns.
- Check for remaining missing values.
- Check for duplicate records.
- Verify the uniqueness of column names.
- Check the data types of important fields.
- Preview the final dataset.

In [24]:
# ============================================================
# STEP 9 — DATA VALIDATION AND QUALITY CHECK
# ============================================================

print("=" * 60)
print("FINAL DATA VALIDATION")
print("=" * 60)

# ------------------------------------------------------------
# 1. Dataset dimensions
# ------------------------------------------------------------

print("\n1. Dataset Dimensions")
print("Rows    :", clean_df.shape[0])
print("Columns :", clean_df.shape[1])


# ------------------------------------------------------------
# 2. Missing value check
# ------------------------------------------------------------

total_missing = clean_df.isna().sum().sum()

print("\n2. Missing Value Check")
print("Total Missing Values:", total_missing)

if total_missing == 0:
    print("Status: PASS — No missing values found.")
else:
    print("Status: REVIEW — Missing values are present.")


# ------------------------------------------------------------
# 3. Duplicate check
# ------------------------------------------------------------

duplicate_count = clean_df.duplicated().sum()

print("\n3. Duplicate Check")
print("Duplicate Rows:", duplicate_count)

if duplicate_count == 0:
    print("Status: PASS — No duplicate rows found.")
else:
    print("Status: REVIEW — Duplicate rows are present.")


# ------------------------------------------------------------
# 4. Column name uniqueness
# ------------------------------------------------------------

unique_columns = clean_df.columns.is_unique

print("\n4. Column Name Check")
print("All Column Names Unique:", unique_columns)

if unique_columns:
    print("Status: PASS")
else:
    print("Status: REVIEW — Duplicate column names found.")


# ------------------------------------------------------------
# 5. Important field checks
# ------------------------------------------------------------

print("\n5. Key Field Check")

key_columns = ["Ep. No.", "Pitch No.", "Brand", "Idea", "City"]

for column in key_columns:
    if column in clean_df.columns:
        print(f"{column}: Available")


# ------------------------------------------------------------
# 6. Data types
# ------------------------------------------------------------

print("\n6. Data Types")
display(clean_df.dtypes.to_frame("Data Type"))


# ------------------------------------------------------------
# 7. Final preview
# ------------------------------------------------------------

print("\n7. Final Dataset Preview")
display(clean_df.head(10))


# ------------------------------------------------------------
# 8. Validation summary
# ------------------------------------------------------------

print("\n" + "=" * 60)
print("VALIDATION SUMMARY")
print("=" * 60)

print("Rows:", clean_df.shape[0])
print("Columns:", clean_df.shape[1])
print("Missing Values:", total_missing)
print("Duplicate Rows:", duplicate_count)
print("Unique Column Names:", unique_columns)

FINAL DATA VALIDATION

1. Dataset Dimensions
Rows    : 124
Columns : 15

2. Missing Value Check
Total Missing Values: 22
Status: REVIEW — Missing values are present.

3. Duplicate Check
Duplicate Rows: 0
Status: PASS — No duplicate rows found.

4. Column Name Check
All Column Names Unique: True
Status: PASS

5. Key Field Check
Ep. No.: Available
Brand: Available
Idea: Available
City: Available

6. Data Types


,Data Type
Ep. No.,Int64
Ep. No..1,object
Pitch No.l,object
Brand,string[python]
Idea,string[python]
City,string[python]
Original ask,string[python]
Deal,string[python]
Anupam Mittal,string[python]
Aman Gupta,string[python]



7. Final Dataset Preview


,Ep. No.,Ep. No..1,Pitch No.l,Brand,Idea,City,Original ask,Deal,Anupam Mittal,Aman Gupta,Ashneer Grover,Namita Thapar,Peyush Bansal,Vineeta Singh,Ghazal Alagh
0,1,1,1,BluePine Industries,Frozen Momos,Faridabad,₹50 lakhs for 5% equity,₹75 lakhs for 16% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
1,1,1,2,Booz scooters,Renting e-bike for mobility in private spaces,Ahmedabad,₹40 lakhs for 15% equity,₹40 lakhs for 50% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
2,1,1,3,Heart up my Sleeves,Detachable Sleeves,New Delhi,₹25 lakhs for 10% equity,₹25 lakhs for 30% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
3,2,2,4,Tagz Foods,Healthy Potato Chips,Bengaluru,₹70 lakhs for 1% equity,₹70 lakhs for 2.75% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
4,2,2,5,Head and Heart,Brain Development Course,Not Available,₹50 lakhs for 5% equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
5,2,2,6,Agri Tourism,Rural Tourism,Not Available,₹50 lakhs for 5% equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
6,3,3,7,qZense Labs,Food Freshness Detector,Bengaluru,₹1 crore for 0.25% equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
7,3,3,8,Peeschute,Disposable Urine Bag,Jalna,₹75 lakhs for 4% equity,₹75 lakhs for 6% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
8,3,3,9,NOCD,Energy Drink,Bengaluru,₹50 lakhs for 2% equity,₹20 lakhs for 15% equity and ₹30 lakhs debt,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
9,4,4,10,CosIQ,Intelligent Skincare,Delhi,₹50 lakhs for 7% equity,₹50 lakhs for 25% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a



VALIDATION SUMMARY
Rows: 124
Columns: 15
Missing Values: 22
Duplicate Rows: 0
Unique Column Names: True


## 9.1 Detailed Missing Value Analysis

The remaining missing values are examined at the column level to identify
where they occur and determine the appropriate treatment without altering
valid source information.

In [25]:
# ============================================================
# STEP 9.1 — DETAILED REMAINING MISSING VALUE ANALYSIS
# ============================================================

# Count missing values in every column
remaining_missing = pd.DataFrame({
    "Column": clean_df.columns,
    "Missing Values": clean_df.isna().sum().values,
    "Missing Percentage": (
        clean_df.isna().mean().values * 100
    ).round(2)
})

# Keep only columns with missing values
remaining_missing = (
    remaining_missing[
        remaining_missing["Missing Values"] > 0
    ]
    .sort_values("Missing Values", ascending=False)
    .reset_index(drop=True)
)

print("Remaining Missing Values by Column:")
display(remaining_missing)

# Display the actual rows containing missing values
print("\nRows Containing Remaining Missing Values:")
display(
    clean_df[
        clean_df.isna().any(axis=1)
    ]
)

Remaining Missing Values by Column:


,Column,Missing Values,Missing Percentage
0,Ep. No.,22,17.74



Rows Containing Remaining Missing Values:


,Ep. No.,Ep. No..1,Pitch No.l,Brand,Idea,City,Original ask,Deal,Anupam Mittal,Aman Gupta,Ashneer Grover,Namita Thapar,Peyush Bansal,Vineeta Singh,Ghazal Alagh
98,<NA>,31,99,Scrapshala,Handmade Reusable Scrap Materials,Not Available,₹50 lakhs for 10% equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
99,<NA>,31,100,Sabjikothi,Vegetables Storage,Not Available,₹60 lakhs for 2.5% equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
100,<NA>,31,101,AyuRythm,Ayurvedic Wellness App,Not Available,₹75 lakhs for 2% Equity,₹75 lakhs for 2.68% Equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
101,<NA>,31,102,Astrix,Smart Locks,Not Available,₹75 lakhs for 3% Equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
102,<NA>,32,103,Thea and Sid,Proposal Solutions,Not Available,₹80 lakhs for 7% Equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,—N/a
103,<NA>,32,104,Experential Etc,Technology layered Advertisement Services,Not Available,₹2 crores for 4% Equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,—N/a
104,<NA>,32,105,Grow fitter,Rewards App,Not Available,₹50 lakhs for 1% Equity,₹50 lakhs for 2% Equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,—N/a
105,<NA>,32,106,Med Tech,Portable ophthalmic devices,Not Available,₹35 lakhs for 6% Equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,—N/a
106,<NA>,33,107,Colour Me Mad,Insoles,Not Available,₹40 lakhs for 10% Equity,₹40 lakhs for 25% Equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,—N/a
107,<NA>,33,108,Mavi's,Vegan Fermented Food,Not Available,₹40 lakhs for 5% Equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,—N/a


### Episode Number Validation

The episode numbers are forward-filled to account for merged cells in the
source HTML table. The cleaned episode and pitch information is then
reviewed using the actual column names generated during scraping.

In [28]:
# ============================================================
# STEP 9.2 — VERIFY EPISODE NUMBER TREATMENT
# ============================================================

# Display the actual column names generated after scraping
print("Available Columns:")
print(clean_df.columns.tolist())

# Find the pitch-number column automatically
pitch_columns = [
    column
    for column in clean_df.columns
    if "Pitch" in str(column)
]

print("\nPitch-related columns found:")
print(pitch_columns)

# Verify episode numbers
print(
    "\nMissing Episode Numbers:",
    clean_df["Ep. No."].isna().sum()
)

# Display the relevant columns using the actual pitch column name
if pitch_columns:
    display(
        clean_df[
            ["Ep. No.", pitch_columns[0], "Brand"]
        ].head(15)
    )
else:
    print("No Pitch Number column was found.")

Available Columns:
['Ep. No.', 'Ep. No..1', 'Pitch No.l', 'Brand', 'Idea', 'City', 'Original ask', 'Deal', 'Anupam Mittal', 'Aman Gupta', 'Ashneer Grover', 'Namita Thapar', 'Peyush Bansal', 'Vineeta Singh', 'Ghazal Alagh']

Pitch-related columns found:
['Pitch No.l']

Missing Episode Numbers: 0


,Ep. No.,Pitch No.l,Brand
0,1,1,BluePine Industries
1,1,2,Booz scooters
2,1,3,Heart up my Sleeves
3,2,4,Tagz Foods
4,2,5,Head and Heart
5,2,6,Agri Tourism
6,3,7,qZense Labs
7,3,8,Peeschute
8,3,9,NOCD
9,4,10,CosIQ


## 9.3 Final Data Quality Validation

The dataset is validated after completing the missing-value treatment and
data-cleaning process.

The final checks confirm that:

- All episode numbers are available.
- No duplicate records remain.
- No missing values remain.
- Column names are unique.
- The dataset has the expected number of rows and columns.

In [29]:
# ============================================================
# STEP 9.3 — FINAL DATA QUALITY VALIDATION
# ============================================================

print("=" * 60)
print("FINAL DATA QUALITY VALIDATION")
print("=" * 60)

# Dataset dimensions
print("\nDataset Dimensions:")
print("Rows    :", clean_df.shape[0])
print("Columns :", clean_df.shape[1])

# Missing values
total_missing = clean_df.isna().sum().sum()

print("\nMissing Values:", total_missing)

# Duplicate records
duplicate_count = clean_df.duplicated().sum()

print("Duplicate Rows:", duplicate_count)

# Column uniqueness
print("Unique Column Names:", clean_df.columns.is_unique)

# Episode number validation
missing_episode_numbers = clean_df["Ep. No."].isna().sum()

print("Missing Episode Numbers:", missing_episode_numbers)

# Final status
if (
    total_missing == 0
    and duplicate_count == 0
    and clean_df.columns.is_unique
    and missing_episode_numbers == 0
):
    print("\nPASS — Dataset passed all final quality checks.")
else:
    print("\nREVIEW — Some quality checks still require attention.")

# Final preview
print("\nFinal Dataset Preview:")
display(clean_df.head(10))

FINAL DATA QUALITY VALIDATION

Dataset Dimensions:
Rows    : 124
Columns : 15

Missing Values: 0
Duplicate Rows: 0
Unique Column Names: True
Missing Episode Numbers: 0

PASS — Dataset passed all final quality checks.

Final Dataset Preview:


,Ep. No.,Ep. No..1,Pitch No.l,Brand,Idea,City,Original ask,Deal,Anupam Mittal,Aman Gupta,Ashneer Grover,Namita Thapar,Peyush Bansal,Vineeta Singh,Ghazal Alagh
0,1,1,1,BluePine Industries,Frozen Momos,Faridabad,₹50 lakhs for 5% equity,₹75 lakhs for 16% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
1,1,1,2,Booz scooters,Renting e-bike for mobility in private spaces,Ahmedabad,₹40 lakhs for 15% equity,₹40 lakhs for 50% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
2,1,1,3,Heart up my Sleeves,Detachable Sleeves,New Delhi,₹25 lakhs for 10% equity,₹25 lakhs for 30% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
3,2,2,4,Tagz Foods,Healthy Potato Chips,Bengaluru,₹70 lakhs for 1% equity,₹70 lakhs for 2.75% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
4,2,2,5,Head and Heart,Brain Development Course,Not Available,₹50 lakhs for 5% equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
5,2,2,6,Agri Tourism,Rural Tourism,Not Available,₹50 lakhs for 5% equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
6,3,3,7,qZense Labs,Food Freshness Detector,Bengaluru,₹1 crore for 0.25% equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
7,3,3,8,Peeschute,Disposable Urine Bag,Jalna,₹75 lakhs for 4% equity,₹75 lakhs for 6% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
8,3,3,9,NOCD,Energy Drink,Bengaluru,₹50 lakhs for 2% equity,₹20 lakhs for 15% equity and ₹30 lakhs debt,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
9,4,4,10,CosIQ,Intelligent Skincare,Delhi,₹50 lakhs for 7% equity,₹50 lakhs for 25% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a


## 10. Export the Final Dataset as CSV

After completing the data cleaning, transformation, and validation process,
the final Shark Tank India dataset is exported as a CSV file.

The exported file is prepared for further analysis, Kaggle publication,
and dataset sharing.

In [30]:
# ============================================================
# STEP 10 — EXPORT FINAL DATASET AS CSV
# ============================================================

# Final output file name
output_file = "shark_tank_india_dataset.csv"

# Export dataset
clean_df.to_csv(
    output_file,
    index=False,
    encoding="utf-8-sig"
)

print("=" * 60)
print("CSV EXPORT COMPLETED")
print("=" * 60)

print("\nFile Name:", output_file)
print("Rows:", clean_df.shape[0])
print("Columns:", clean_df.shape[1])

# Verify exported file by reading it back
exported_df = pd.read_csv(output_file)

print("\nExport Verification:")
print("Rows:", exported_df.shape[0])
print("Columns:", exported_df.shape[1])

print("\nFirst 5 Rows of Exported Dataset:")
display(exported_df.head())

CSV EXPORT COMPLETED

File Name: shark_tank_india_dataset.csv
Rows: 124
Columns: 15

Export Verification:
Rows: 124
Columns: 15

First 5 Rows of Exported Dataset:


,Ep. No.,Ep. No..1,Pitch No.l,Brand,Idea,City,Original ask,Deal,Anupam Mittal,Aman Gupta,Ashneer Grover,Namita Thapar,Peyush Bansal,Vineeta Singh,Ghazal Alagh
0,1,1,1,BluePine Industries,Frozen Momos,Faridabad,₹50 lakhs for 5% equity,₹75 lakhs for 16% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
1,1,1,2,Booz scooters,Renting e-bike for mobility in private spaces,Ahmedabad,₹40 lakhs for 15% equity,₹40 lakhs for 50% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
2,1,1,3,Heart up my Sleeves,Detachable Sleeves,New Delhi,₹25 lakhs for 10% equity,₹25 lakhs for 30% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
3,2,2,4,Tagz Foods,Healthy Potato Chips,Bengaluru,₹70 lakhs for 1% equity,₹70 lakhs for 2.75% equity,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a
4,2,2,5,Head and Heart,Brain Development Course,Not Available,₹50 lakhs for 5% equity,No Deal,No Investment Recorded,No Investment Recorded,No Investment Recorded,No Investment Recorded,—N/a,No Investment Recorded,—N/a


## 11. Dataset Name

### Shark Tank India Season 1 — Pitches & Investments Dataset

The dataset name is designed to be descriptive and search-friendly. It clearly
identifies the source context, season, and the primary information contained
in the dataset.


## 12. Kaggle Dataset Publication

The cleaned and validated dataset has been exported as a CSV file and is
ready to be published on Kaggle.

The dataset will be published on Kaggle using the following details:

- **Dataset Name:** Shark Tank India Season 1 — Pitches & Investments Dataset
- **File:** shark_tank_india_dataset.csv
- **Rows:** 124
- **Columns:** 15
- **Format:** CSV
- **Source:** Shark Tank India Season 1 webpage
- **Data Collection Method:** Web scraping using `pandas.read_html()`
- **Data Processing:** Data cleaning, missing-value treatment, duplicate
  removal, data type optimization, and validation